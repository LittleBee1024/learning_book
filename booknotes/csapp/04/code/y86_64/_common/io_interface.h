#pragma once

#include <stdio.h>

namespace IO
{
   class InputInterface
   {
   public:
      /**
       * @brief Get the handler of flex input
       */
      virtual FILE *getHandler() = 0;

   public:
      InputInterface() = default;
      InputInterface(const InputInterface &) = delete;
      InputInterface &operator=(const InputInterface &) = delete;
      virtual ~InputInterface() = default;
   };

   class OutputInterface
   {
   public:
      /**
       * @brief Output string
       *
       * @param format The format of the string
       */
      virtual void out(const char *format, ...) = 0;

   public:
      OutputInterface() = default;
      OutputInterface(const OutputInterface &) = delete;
      OutputInterface &operator=(const OutputInterface &) = delete;
      virtual ~OutputInterface() = default;
   };
}
