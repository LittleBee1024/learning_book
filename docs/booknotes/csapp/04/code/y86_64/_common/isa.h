#pragma once

#include <cstddef>

typedef unsigned char byte_t;
typedef long long int word_t;
typedef long long unsigned uword_t;

/******************
 * Y86_64 Registers
 ******************/
/* REG_NONE is a special one to indicate no register */
enum REG_ID : int
{
   REG_RAX,
   REG_RCX,
   REG_RDX,
   REG_RBX,
   REG_RSP,
   REG_RBP,
   REG_RSI,
   REG_RDI,
   REG_R8,
   REG_R9,
   REG_R10,
   REG_R11,
   REG_R12,
   REG_R13,
   REG_R14,
   REG_NONE = 0xF,
   REG_ERR
};

#define VALID_REG(id) (id >= 0 && id < REG_NONE)

/******************
 * Y86_64 Instructions
 ******************/
/* Different instruction types */
enum INSTR : int
{
   I_HALT,
   I_NOP,
   I_RRMOVQ,
   I_IRMOVQ,
   I_RMMOVQ,
   I_MRMOVQ,
   I_ALU,
   I_JMP,
   I_CALL,
   I_RET,
   I_PUSHQ,
   I_POPQ,
   I_IADDQ,
   I_POP2
};

/* Default function code */
enum I_FUN : int
{
   F_NONE
};

/* Different instruction argument types */
enum INSTR_ARG : int
{
   R_ARG,
   M_ARG,
   I_ARG,
   NO_ARG
};

/* Different ALU operations */
enum ALU : int
{
   A_ADD,
   A_SUB,
   A_AND,
   A_XOR,
   A_NONE
};

/* Different Jump conditions */
enum COND : int
{
   C_YES,
   C_LE,
   C_L,
   C_E,
   C_NE,
   C_GE,
   C_G
};

/* Pack itype and function into single byte */
#define HPACK(hi, lo) ((((hi)&0xF) << 4) | ((lo)&0xF))

/* Unpack byte */
#define HI4(byte) (((byte) >> 4) & 0xF)
#define LO4(byte) ((byte)&0xF)

/* Get the opcode out of one byte instruction field */
#define GET_ICODE(instr) HI4(instr)

/* Get the ALU/JMP function out of one byte instruction field */
#define GET_FUN(instr) LO4(instr)

/******************
 * Y86_64 Condition
 ******************/
typedef unsigned char cc_t;

#define GET_ZF(cc) (((cc) >> 2) & 0x1)
#define GET_SF(cc) (((cc) >> 1) & 0x1)
#define GET_OF(cc) (((cc) >> 0) & 0x1)

#define PACK_CC(z, s, o) (((z) << 2) | ((s) << 1) | ((o) << 0))
#define DEFAULT_CC PACK_CC(1, 0, 0)

namespace ISA
{
   /**
    * @brief Y86-64 instruction structure
    */
   struct Instr
   {
      const char *name;
      unsigned char code; /* Byte code for instruction+op */
      int bytes;
      INSTR_ARG arg1;
      int arg1pos;
      int arg1hi; /* 0/1 for register argument, # bytes for allocation */
      INSTR_ARG arg2;
      int arg2pos;
      int arg2hi; /* 0/1 */
   };

   /**
    * @brief Find Y86-64 instruction struction by the instruction name
    * 
    * @param name Instruction name
    * @return const Instr&, return the reference of Y86-64 instruction struction.
    *         If no instruction is found, return the empty instruction.
    */
   const Instr &findInstr(const char *name);

   /**
    * @brief Decode Y86-64 instruction name by the icode
    * 
    * @param code Instruction code number
    * @return const char*, return the instruction name. For invalid icode, return nullptr.
    */
   const char *decodeInstrName(int code);

   /**
    * @brief Find Y86-64 reigster ID by register name
    * 
    * @param name Register name
    * @return Y_REG_ID, for invalid register name, return Y_REG_ID::ERR
    */
   REG_ID findRegister(const char *name);

   /**
    * @brief Get the Y86-64 register name by register ID
    * 
    * @param id Register ID
    * @return const char*, for invalid ID, return nullptr
    */
   const char *getRegName(REG_ID id);

   /**
    * @brief Get the name of Y76-64 CC
    * 
    * @param c, the type of CC
    * @return const char* 
    */
   const char *getCCName(cc_t c);
}
