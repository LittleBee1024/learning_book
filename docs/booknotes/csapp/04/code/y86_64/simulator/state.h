#pragma once

#include <cstdint>

namespace SIM
{
   enum State : int
   {
      STAT_OK,
      STAT_HLT,
      STAT_BUBBLE,
      STAT_ERR_ADDR,
      STAT_ERR_REG,
      STAT_ERR_INSTR,
      STAT_ERR_PIPE
   };

   const char *getStateName(State s);
}
