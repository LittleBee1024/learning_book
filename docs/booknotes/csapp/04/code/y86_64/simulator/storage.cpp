#include "./simulator/storage.h"

#include <assert.h>
#include <string>

namespace
{
   byte_t toNum(char ch, char cl)
   {
      std::string hexNum;
      hexNum.push_back(ch);
      hexNum.push_back(cl);
      return std::stoi(hexNum, nullptr, 16);
   }
}

namespace SIM
{
   Storage::Storage(int len, std::shared_ptr<IO::OutputInterface> out) : m_out(out)
   {
      len = (len + SIZE_ALIGN - 1) / SIZE_ALIGN * SIZE_ALIGN;
      m_contents.resize(len, 0);
   }

   bool Storage::getByte(word_t pos, byte_t *dest) const
   {
      if (pos < 0 || pos >= (word_t)m_contents.size())
         return false;
      *dest = m_contents[pos];
      return true;
   }

   bool Storage::setByte(word_t pos, byte_t val)
   {
      if (pos < 0 || pos >= (word_t)m_contents.size())
         return false;
      m_contents[pos] = val;
      return true;
   }

   bool Storage::getWord(word_t pos, word_t *dest) const
   {
      if (pos < 0 || pos + sizeof(word_t) >= m_contents.size())
         return false;

      word_t val = 0;
      for (size_t i = 0; i < sizeof(word_t); i++)
      {
         // little-endian
         word_t b = m_contents[pos + i] & 0xFF;
         val = val | (b << (sizeof(word_t) * i));
      }
      *dest = val;
      return true;
   }

   bool Storage::setWord(word_t pos, word_t val)
   {
      if (pos < 0 || pos + sizeof(word_t) > m_contents.size())
         return false;
      for (size_t i = 0; i < sizeof(word_t); i++)
      {
         m_contents[pos + i] = (byte_t)val & 0xFF;
         val >>= sizeof(word_t);
      }
      return true;
   }

   size_t Storage::size() const
   {
      return m_contents.size();
   }

   void Storage::reset()
   {
      std::fill(m_contents.begin(), m_contents.end(), 0);
   }

   RegStore::RegStore(int len, std::shared_ptr<IO::OutputInterface> out) : Storage(len, out)
   {
   }

   word_t RegStore::getRegVal(REG_ID id) const
   {
      word_t val = 0;
      if (id >= REG_NONE)
         return 0;
      getWord(id * sizeof(word_t), &val);
      return val;
   }

   void RegStore::setRegVal(REG_ID id, word_t val)
   {
      if (id < REG_NONE)
      {
         setWord(id * sizeof(word_t), val);
      }
   }

   MemStore::MemStore(int len, std::shared_ptr<IO::OutputInterface> out) : Storage(len, out)
   {
   }

   void MemStore::loadOneInstr(int lineno, word_t addr, std::string instr)
   {
      for (size_t i = 0; i < instr.size(); i += 2)
      {
         if (addr >= (word_t)m_contents.size())
         {
            m_out->out("[ERROR] Invalid addr (Line %d): 0x%llx\n", lineno, addr);
            return;
         }
         // little-endian, the least-significant byte at the smallest address
         m_contents[addr++] = toNum(instr[i], instr[i + 1]);
      }
   }
}
