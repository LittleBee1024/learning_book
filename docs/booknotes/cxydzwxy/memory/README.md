# 内存

> [《程序员的自我修养--链接、装载与库》 - 俞甲子，石凡，潘爱民](https://1drv.ms/b/s!AkcJSyT7tq80cUuHb2eRcJkkBjM?e=YUwBqB)

## 分段(Segmentation)
## 分页(Paging)
