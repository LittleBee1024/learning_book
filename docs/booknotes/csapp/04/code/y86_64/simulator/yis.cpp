#include "./yis.h"
#include "isa.h"
#include <assert.h>

namespace SIM
{
   YIS::YIS(IO::OutputInterface &out) : m_out(out),
                                        m_pc(0),
                                        m_reg(REG_SIZE_BYTES, m_out),
                                        m_mem(MEM_SIZE_BYTES, m_out)
   {
   }

   int YIS::loadCode(const char *fname)
   {
      int bytes = m_mem.load(fname);
      if (bytes)
         m_out.out("[INFO] Loaded %d bytes code\n", bytes);
      return bytes;
   }

   State YIS::runOneStep()
   {
      word_t ftpc = m_pc; // fall-through PC
      byte_t byte0 = 0;   // (icode+ifun)
      if (!m_mem.getByte(ftpc, &byte0))
      {
         m_out.out("[ERROR] PC = 0x%llx, Invalid instruction address\n", m_pc);
         return STAT_ERR_ADDR;
      }
      ftpc++;

      INSTR icode = (INSTR)HI4(byte0);
      I_FUN ifun = (I_FUN)LO4(byte0);

      byte_t byte1 = 0; // (rA+rB)
      bool needReg = (icode == I_RRMOVQ || icode == I_ALU || icode == I_PUSHQ ||
                      icode == I_POPQ || icode == I_IRMOVQ || icode == I_RMMOVQ ||
                      icode == I_MRMOVQ || icode == I_IADDQ);
      REG_ID rA = REG_NONE;
      REG_ID rB = REG_NONE;
      if (needReg)
      {
         if (!m_mem.getByte(ftpc, &byte1))
         {
            m_out.out("[ERROR] PC = 0x%llx, Invalid instruction address\n", m_pc);
            return STAT_ERR_ADDR;
         }
         ftpc++;

         if (icode == I_RRMOVQ || icode == I_ALU || icode == I_RMMOVQ || icode == I_MRMOVQ)
         {
            rA = (REG_ID)HI4(byte1);
            rB = (REG_ID)LO4(byte1);
            if (!VALID_REG(rA) || !VALID_REG(rB))
            {
               m_out.out("[ERROR] PC = 0x%llx, Invalid registers rA=0x%.1x, rB=0x%.1x\n", m_pc, rA, rB);
               return STAT_ERR_REG;
            }
         }

         if (icode == I_PUSHQ || icode == I_POPQ)
         {
            rA = (REG_ID)HI4(byte1);
            if (!VALID_REG(rA))
            {
               m_out.out("[ERROR] PC = 0x%llx, Invalid register rA=0x%.1x\n", m_pc, rA);
               return STAT_ERR_REG;
            }
         }

         if (icode == I_IRMOVQ || icode == I_IADDQ)
         {
            rB = (REG_ID)LO4(byte1);
            if (!VALID_REG(rB))
            {
               m_out.out("[ERROR] PC = 0x%llx, Invalid register rB=0x%.1x\n", m_pc, rB);
               return STAT_ERR_REG;
            }
         }
      }

      word_t cval = 0;
      bool needImm = (icode == I_IRMOVQ || icode == I_RMMOVQ || icode == I_MRMOVQ ||
                      icode == I_JMP || icode == I_CALL || icode == I_IADDQ);
      if (needImm)
      {
         if (!m_mem.getWord(ftpc, &cval))
         {
            m_out.out("[ERROR] PC = 0x%llx, Invalid instruction address\n", m_pc);
            return STAT_ERR_ADDR;
         }
         ftpc += sizeof(word_t);
      }

      switch (icode)
      {
      case I_HALT:
         return STAT_HLT;
      case I_NOP:
         m_pc = ftpc;
         break;
      case I_RRMOVQ: // rrmovq rA, rB || cmovXX rA, rB
      {
         word_t val = m_reg.getRegVal(rA);
         if (checkCond(m_cc, (COND)ifun))
            m_reg.setRegVal(rB, val);
         m_pc = ftpc;
         break;
      }
      case I_IRMOVQ: // irmovq V, rB
      {
         m_reg.setRegVal(rB, cval);
         m_pc = ftpc;
         break;
      }
      case I_RMMOVQ: // rmmovq rA, D(rB)
      {
         cval += m_reg.getRegVal(rB);
         word_t val = m_reg.getRegVal(rA);
         if (m_mem.setWord(cval, val))
         {
            m_out.out("[ERROR] PC = 0x%llx, Invalid RMMOVQ instruction address: 0x%llx\n", m_pc, cval);
            return STAT_ERR_ADDR;
         }
         m_pc = ftpc;
         break;
      }
      case I_MRMOVQ: // mrmovq D(rB), rA
      {
         cval += m_reg.getRegVal(rB);
         word_t val = 0;
         if (!m_mem.getWord(cval, &val))
         {
            m_out.out("[ERROR] PC = 0x%llx, Invalid MRMOVQ instruction address: 0x%llx\n", m_pc, cval);
            return STAT_ERR_ADDR;
         }
         m_reg.setRegVal(rA, val);
         m_pc = ftpc;
         break;
      }
      case I_ALU: // OPq rA, rB
      {
         word_t argA = m_reg.getRegVal(rA);
         word_t argB = m_reg.getRegVal(rB);
         word_t val = computeALU((ALU)ifun, argA, argB);
         m_reg.setRegVal(rB, val);
         m_cc = computeCC((ALU)ifun, argA, argB);
         m_pc = ftpc;
         break;
      }
      case I_JMP: // jXX Dest
      {
         if (checkCond(m_cc, (COND)ifun))
            m_pc = cval;
         else
            m_pc = ftpc;
         break;
      }
      case I_CALL: // call Dest
      {
         word_t val = m_reg.getRegVal(REG_RSP) - sizeof(word_t);
         m_reg.setRegVal(REG_RSP, val);
         // push PC to stack
         if (!m_mem.setWord(val, ftpc)) // write address to stack
         {
            m_out.out("[ERROR] PC = 0x%llx, Invalid CALL instruction stack address: 0x%llx\n", m_pc, val);
            return STAT_ERR_ADDR;
         }
         m_pc = cval;
         break;
      }
      case I_RET: // ret
      {
         // pop address from stack
         word_t dval = m_reg.getRegVal(REG_RSP);
         word_t val = 0;
         if (!m_mem.getWord(dval, &val)) // read from stack
         {
            m_out.out("[ERROR] PC = 0x%llx, Invalid RET instruction stack address: 0x%llx\n", m_pc, dval);
            return STAT_ERR_ADDR;
         }
         m_reg.setRegVal(REG_RSP, dval + sizeof(word_t));
         m_pc = val;
         break;
      }
      case I_PUSHQ: // pushq rA
      {
         word_t val = m_reg.getRegVal(rA);
         word_t dval = m_reg.getRegVal(REG_RSP) - sizeof(word_t);
         m_reg.setRegVal(REG_RSP, dval);
         if (!m_mem.setWord(dval, val))
         {
            m_out.out("[ERROR] PC = 0x%llx, Invalid PUSHQ instruction stack address: 0x%llx\n", m_pc, dval);
            return STAT_ERR_ADDR;
         }
         m_pc = ftpc;
         break;
      }
      case I_POPQ: // popq rA
      {
         word_t dval = m_reg.getRegVal(REG_RSP);
         m_reg.setRegVal(REG_RSP, dval + sizeof(word_t));
         word_t val = 0;
         if (!m_mem.getWord(dval, &val))
         {
            m_out.out("[ERROR] PC = 0x%llx, Invalid POPQ instruction stack address: 0x%llx\n", m_pc, dval);
            return STAT_ERR_ADDR;
         }
         m_reg.setRegVal(rA, val);
         m_pc = ftpc;
         break;
      }
      case I_IADDQ:
      {
         word_t argB = m_reg.getRegVal(rB);
         word_t val = argB + cval;
         m_reg.setRegVal(rB, val);
         m_cc = computeCC(A_ADD, cval, argB);
         m_pc = ftpc;
         break;
      }
      default:
      {
         m_out.out("[ERROR] PC = 0x%llx, Invalid instruction %.2x\n", m_pc, byte0);
         return STAT_ERR_INSTR;
      }
      }
      return STAT_OK;
   }

   void YIS::compare(const SimInterface &other) const
   {
      const YIS *child = dynamic_cast<const YIS *>(&other);
      if (!child)
      {
         m_out.out("[ERROR] Compared with an invalid simulator snapshot\n");
         return;
      }

      compareReg(*child);
      compareMem(*child);
   }

   void YIS::compareReg(const YIS &other) const
   {
      m_out.out("Changes to registers:\n");

      for (int id = REG_RAX; id < REG_NONE; id++)
      {
         word_t oldVal = other.m_reg.getRegVal((REG_ID)id);
         word_t newVal = m_reg.getRegVal((REG_ID)id);
         if (oldVal != newVal)
         {
            m_out.out("%s:\t0x%.16llx\t0x%.16llx\n", ISA::getRegName((REG_ID)id), oldVal, newVal);
         }
      }
   }

   void YIS::compareMem(const YIS &other) const
   {
      m_out.out("Changes to memory:\n");

      assert(m_mem.size() == other.m_mem.size());
      for (size_t i = 0; i < m_mem.size(); i += sizeof(word_t))
      {
         word_t oldVal = 0;
         other.m_mem.getWord(i, &oldVal);
         word_t newVal = 0;
         m_mem.getWord(i, &newVal);
         if (oldVal != newVal)
         {
            m_out.out("0x%.4llx:\t0x%.16llx\t0x%.16llx\n", i, oldVal, newVal);
         }
      }
   }
}
