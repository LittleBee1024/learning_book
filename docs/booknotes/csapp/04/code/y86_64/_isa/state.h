#pragma once

#include "io_interface.h"
#include "./def.h"
#include "./storage.h"

#include <memory>

namespace ISA
{

   enum class StateType
   {
      OK,
      HALT,
      INVALID_ADDR,
      INVALID_REG,
      INVALID_INSTR
   };

   class State
   {
      static constexpr int REG_SIZE_BYTES = 128;       // 8 bytes * 16 regs
      static constexpr int MEM_SIZE_BYTES = (1 << 13); // 8KB
   public:
      explicit State(std::unique_ptr<CO::OutputInterface> &&out);
      int loadCode(const char *fname);
      StateType step();

   private:
      bool checkCond(COND c);
      word_t computeALU(ALU op, word_t argA, word_t argB);
      cc_t computeCC(ALU op, word_t argA, word_t argB);

   private:
      std::unique_ptr<CO::OutputInterface> m_out;

      word_t m_pc;
      RegStore m_reg;
      MemStore m_mem;
      cc_t m_cc;
   };

}
