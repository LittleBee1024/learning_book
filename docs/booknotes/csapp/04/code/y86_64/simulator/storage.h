#pragma once

#include "./_common/io_interface.h"
#include "./_common/isa.h"

#include <memory>
#include <vector>

namespace SIM
{
   class Storage
   {
      static constexpr int SIZE_ALIGN = 32;
   public:
      Storage(int len, std::shared_ptr<IO::OutputInterface> out);
      Storage(const Storage &other) = default;
      Storage &operator=(const Storage &) = default;
      virtual ~Storage() = default;

      bool getByte(word_t pos, byte_t *dest) const;
      bool setByte(word_t pos, byte_t val);

      bool getWord(word_t pos, word_t *dest) const;
      bool setWord(word_t pos, word_t val);

      size_t size() const;
      void reset();

   protected:
      std::shared_ptr<IO::OutputInterface> m_out;
      std::vector<byte_t> m_contents;
   };

   class MemStore : public Storage
   {
   public:
      MemStore(int len, std::shared_ptr<IO::OutputInterface> out);
      void loadOneInstr(int lineno, word_t addr, std::string instr);
   };

   class RegStore : public Storage
   {
   public:
      RegStore(int len, std::shared_ptr<IO::OutputInterface> out);

      word_t getRegVal(REG_ID id) const;
      void setRegVal(REG_ID id, word_t val);
   };
}