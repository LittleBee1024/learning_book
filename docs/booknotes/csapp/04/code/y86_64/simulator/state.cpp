#include "./state.h"

namespace SIM
{
   const char *getStateName(State s)
   {
      static const char *stateNames[] = { "OK", "HALT", "ERR_ADR", "ERR_REG", "ERR_INSTR" };
      return stateNames[s];
   }
}
