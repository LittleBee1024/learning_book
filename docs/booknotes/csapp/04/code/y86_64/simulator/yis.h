#pragma once

#include "./state.h"
#include "./storage.h"

#include "io_interface.h"
#include "isa.h"

#include <memory>

namespace SIM
{

   class YIS
   {
      static constexpr int REG_SIZE_BYTES = 128;       // 8 bytes * 16 regs
      static constexpr int MEM_SIZE_BYTES = (1 << 13); // 8KB
   public:
      explicit YIS(std::unique_ptr<IO::OutputInterface> &&out);
      int loadCode(const char *fname);
      State step();

   private:
      bool checkCond(COND c);
      word_t computeALU(ALU op, word_t argA, word_t argB);
      cc_t computeCC(ALU op, word_t argA, word_t argB);

   private:
      std::unique_ptr<IO::OutputInterface> m_out;

      word_t m_pc;
      RegStore m_reg;
      MemStore m_mem;
      cc_t m_cc;
   };

}
