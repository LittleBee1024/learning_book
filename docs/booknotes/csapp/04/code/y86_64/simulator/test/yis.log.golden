[INFO] Create Yis Simulator
[INFO] Loaded 141 bytes code

Cycle 0. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x0
D: instr = irmovq, rA = ----, rB = %rsp, valC = 0x200, valP = 0xa

Cycle 1. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0xa
D: instr = call, rA = ----, rB = ----, valC = 0x38, valP = 0x13

Cycle 2. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x38
D: instr = irmovq, rA = ----, rB = %rdi, valC = 0x18, valP = 0x42

Cycle 3. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x42
D: instr = irmovq, rA = ----, rB = %rsi, valC = 0x4, valP = 0x4c

Cycle 4. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x4c
D: instr = call, rA = ----, rB = ----, valC = 0x56, valP = 0x55

Cycle 5. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x56
D: instr = irmovq, rA = ----, rB = %r8, valC = 0x8, valP = 0x60

Cycle 6. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x60
D: instr = irmovq, rA = ----, rB = %r9, valC = 0x1, valP = 0x6a

Cycle 7. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x6a
D: instr = xorq, rA = %rax, rB = %rax, valC = 0x0, valP = 0x6c

Cycle 8. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x6c
D: instr = andq, rA = %rsi, rB = %rsi, valC = 0x0, valP = 0x6e

Cycle 9. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x6e
D: instr = jmp, rA = ----, rB = ----, valC = 0x87, valP = 0x77

Cycle 10. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90

Cycle 11. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81

Cycle 12. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83

Cycle 13. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85

Cycle 14. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87

Cycle 15. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90

Cycle 16. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81

Cycle 17. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83

Cycle 18. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85

Cycle 19. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87

Cycle 20. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90

Cycle 21. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81

Cycle 22. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83

Cycle 23. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85

Cycle 24. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87

Cycle 25. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90

Cycle 26. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81

Cycle 27. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83

Cycle 28. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85

Cycle 29. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87

Cycle 30. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90

Cycle 31. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x90
D: instr = ret, rA = ----, rB = ----, valC = 0x0, valP = 0x91

Cycle 32. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x55
D: instr = ret, rA = ----, rB = ----, valC = 0x0, valP = 0x56

Cycle 33. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x13
D: instr = halt, rA = ----, rB = ----, valC = 0x0, valP = 0x14

After 34 cycles, the state becomes HALT

Changes to registers:
%rax:	0x0000000000000000	0x0000abcdabcdabcd
%rsp:	0x0000000000000000	0x0000000000000200
%rdi:	0x0000000000000000	0x0000000000000038
%r8:	0x0000000000000000	0x0000000000000008
%r9:	0x0000000000000000	0x0000000000000001
%r10:	0x0000000000000000	0x0000a000a000a000
Changes to memory:
0x01f0:	0x0000000000000000	0x0000000000000055
0x01f8:	0x0000000000000000	0x0000000000000013
